Cypher: document relationship pattern predicates

blouerat opened the pull request:
Note that this depends on https://github.com/neo-technology/neo4j/pull/12784

davidoliverSP2 commented:
Thanks @blouerat - before I review this, could I check if this is for version 4.4?

arnefischereit commented:
No, it is for 5.0.

hedengran reviewed (approved):
Nice! 👍

martin-neotech commented:
@blouerat please rebase

davidoliverSP2 commented:
Hi @blouerat 

I rebased this PR with the latest fixes for the `dev` branch, but it still has a couple of build failures which look relevant to your changes: 

_JUnit Vintage: org.neo4j.cypher.docgen.WhereTest.29:_
`WITH 2000 AS minYear MATCH {a:Person}-[r:KNOWS WHERE r.since < minYear]->{b:Person} RETURN r.since`

_JUnit Vintage: org.neo4j.cypher.docgen.WhereTest.30:_
`WITH 2000 AS minYear MATCH {a:Person {name: 'Andy'}} RETURN [{a}-[r:KNOWS WHERE r.since < minYear]->{b:Person} | r.since] AS years`

Please could you take a look? many thanks

arnefischereit commented:
@davidoliverSP2: The relevant change has since been deemed to not be ready to be surfaced yet, so we put it behind a semantic feature.

I therefore think that this should at the moment not be merged to documentation either. But we might use it for documenting once the feature is fully in place.

WDYT, @blouerat ?

blouerat commented:
Correct, thank you @arnefischereit
I agree that it shouldn't be in the doc until we are ready to remove the semantic feature, we're not there yet.
What's the correct approach, should I close this PR and keep the branch open?

davidoliverSP2 commented:
@blouerat 

I think this PR could be made into a _draft_ PR if it's helpful to keep the content available? Otherwise, yes this can be closed and resubmitted when ready.

blouerat commented:
Sounds good, will turn into a draft PR for the time being. Thank you @davidoliverSP2

blouerat commented:
Re-opening this PR as path pattern predicates have been merged and will be part of 5.0

arnefischereit commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/WhereTest.scala:
```diff
@@ -501,6 +503,27 @@ class WhereTest extends DocumentingTest {
+              r.toList shouldBe List(Map("r.since" -> 1999))
+            }) {
+            resultTable()
+            p("When used this way, predicates in `WHERE` can reference the relationship variable that the `WHERE` clause belongs to, but not other elements of the `MATCH` pattern.")
```
```suggestion
            p("When used this way, predicates in `WHERE` can reference the relationship variable that the `WHERE` clause belongs to, but no other elements of the `MATCH` pattern.")
```

arnefischereit reviewed (changes requested):
I think it makes sense to also add information about the restrictions:
- Predicates are only allowed in name relationships (same for nodes, we may need to add that)
- Predicates are not allowed on variable length relationships or in shortest path (for now)

... and maybe we should express that these are equivalent to writing the WHERE clause outside and will not change behaviour but are just syntactical sugar.

blouerat committed:
Cypher: document relationship pattern predicates

blouerat commented:
I realised that the whole _Pattern element predicates_ was inside of the _Using ranges_ one, which makes no sense, so I've fixed that too.
I've removed the caveats about referencings other variables in both relationship and node patterns as it isn't true anymore.

Here is the rendered version:
![Screen Shot 2022-07-07 at 16 44 01](https://user-images.githubusercontent.com/347039/177802260-a6d3415e-8123-4b57-928e-e1097f171429.png)

davidoliverSP2 commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/WhereTest.scala:
```diff
@@ -467,39 +470,78 @@ class WhereTest extends DocumentingTest {
+          }) {
+          resultTable()
+        }
+        p("However, it cannot be used inside of variable length relationships:")
+        query(
+          """WITH 2000 AS minYear
+            #MATCH (a:Person)-[r:KNOWS*1..3 WHERE r.since > b.yearOfBirth]->(b:Person)
+            #RETURN r.since""".stripMargin('#'),
+          ErrorAssertions(_.getMessage should startWith("Relationship pattern predicates are not supported for variable-length relationships"))
+        )(
+          p("this would lead to an error.")
+        )
```
```suggestion
        p("However, it cannot be used inside of variable length relationships, as this would lead to an error. For example:")
        query(
          """WITH 2000 AS minYear
            #MATCH (a:Person)-[r:KNOWS*1..3 WHERE r.since > b.yearOfBirth]->(b:Person)
            #RETURN r.since""".stripMargin('#'),
          ErrorAssertions(_.getMessage should startWith("Relationship pattern predicates are not supported for variable-length relationships"))
        )
```

davidoliverSP2 replied:
This change removes the hanging sentence after the example.

davidoliverSP2 commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/WhereTest.scala:
```diff
@@ -467,39 +470,78 @@ class WhereTest extends DocumentingTest {
+        )(
+          p("this would lead to an error.")
+        )
+        p("Putting predicates inside a relationship pattern helps with readability, note that it is strictly equivalent to using a standalone `WHERE` sub-clause:")
```
```suggestion
        p("Putting predicates inside a relationship pattern can help with readability. Please note that it is strictly equivalent to using a standalone `WHERE` sub-clause:")
```

davidoliverSP2 reviewed (approved):
Thanks @blouerat 

Just a couple of editorial suggestions to look at

blouerat committed:
Apply suggestions from code review

Co-authored-by: David Oliver <davidoliverSP2@users.noreply.github.com>

blouerat committed:
Add empty innerCodeBlock to query

martin-neotech commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/WhereTest.scala:
```diff
@@ -452,12 +454,13 @@ class WhereTest extends DocumentingTest {
-        ResultAssertions((r) => {
-          r.toList should equal(List(Map("a.name" -> "Timothy", "a.age" -> 25l), Map("a.name" -> "Peter", "a.age" -> 35l)))
-        })) {
+        query(
+          """MATCH (a:Person)
+            #WHERE a.name >= 'Peter'
+            #RETURN a.name, a.age""".stripMargin('#'),
+          ResultAssertions((r) => {
+            r.toList should equal(List(Map("a.name" -> "Timothy", "a.age" -> 25l), Map("a.name" -> "Peter", "a.age" -> 35l)))
+          })) {
```
Not a change...

martin-neotech commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/WhereTest.scala:
```diff
@@ -467,39 +470,76 @@ class WhereTest extends DocumentingTest {
-        ResultAssertions((r) => {
-          r.toList should equal(List(Map("a.name" -> "Peter", "a.age" -> 35l)))
-        })) {
+          ResultAssertions((r) => {
+            r.toList should equal(List(Map("a.name" -> "Peter", "a.age" -> 35l)))
+          })) {
```
Not a change...
